### ISSUE TITLE
Parse file larger than 4 GB

### ISSUE THREAD

**Issue by tahasabih** [simdjson/simdjson#128]
I have JSON dump documents larger than 4GB. Is it possible to modify the library to be able to parse those documents? 

I do want to only parse a subtree of the JSON file, will that be possible without parsing the entire file?

**Comment by lemire** [comment]
Yes... but we don’t have the hooks in the library to do it now. I’d be interested in making it happen if you want to help.

**Comment by tahasabih** [comment]
Yes that would be great!

**Comment by lemire** [comment]
Ok... so what do you imagine as an API?

**Comment by tahasabih** [comment]
I was actually using the python wrapper for this library which provides an interface:
https://github.com/TkTech/pysimdjson -> It's called simdjson.ParsedJson with an example in the README

This is exactly how I want to use it - ideally it only parses the json when it encounters the subtree with a key which matches a regex (or a convention based path).

**Comment by lemire** [comment]
I am not sure pysimdjson does selective parsing. I think it parses the whole thing. 

We enforce the 4GB limit to prevent adversarial inputs where a given string would reach 4GB. We don’t support documents with 4GB strings.


There are other issues with 4GB document like memory usage. If you have very large documents, you probably don’t want to fully materialize the tree as we do currently.

I’m interested in the use case of very large documents but I suspect lifting the recently introduced 4GB limit is the lazy way out. For example, we have done no testing whatsoever on documents that big, no benchmark... And materializing the whole thing may or may not be benificial.

I think the first step would be to do some careful tests and benchmarks for that use case... If you are willing to help out... The 4GB is enforced through a single test and easy to lift. Testing without it is easy. It would be interesting to learn about the performance characteristics of your use case.

**Comment by tahasabih** [comment]
Materializing the whole tree is definitely not the way to go at least for my use case.
In my use-case I have a 6GB nosql json dump from a db and I want to selectively access certain subtrees where I fully know the path to this subtree.

For example:
```
{
	"abc": {
		"def": {
			"ghi": 22
		}
	},
	"qq": {
		"oo": 2
	}
}
```
I just want "abc.def" which should give me
```
{
    "ghi": 22
}
```

I am most definitely willing to help out with benchmarking (though the db dump I have has sensitive data so won't be able to make that document part of this repo for future testing).

I did try removing this line:
```
if(len > SIMDJSON_MAXSIZE_BYTES) {
      return false;
}
```
but the library was still unable to load the data (or maybe I was doing something wrong).

Let me know the next steps! This is most definitely a really useful library and I would love to improve it so I can use it myself as well :)

**Comment by lemire** [comment]
I would not want a 6 GB data file in our repo. nor should we ever need private information.

As far as I know, there is nothing else (except for the line that you found) that should block the use of large (>4GB) file, other than this check. If you use a smaller file (<4GB) or a subset of your existing large file (6GB), does it validate? I'd be interested in knowing more about what prevents you from loading your 6GB file. Well, quite obviously, memory usage is going to be quite high because of the materialization of all this data, and a memory allocation failure could stop the processing (but you'd be able to tell that it happens at that stage).

As for the API, I think that the answer is that you are looking for support for a query language of some sort. This has been raised in issue 95... and it will (or is) being worked on... 
https://github.com/lemire/simdjson/issues/95

Of course, feel free to contribute as well to issue 95.

**Comment by tahasabih** [comment]
Yeah I had a smaller file (~3GB) with the same structure and that was being parsed ok.

I think the memory allocation stage is failing - but I am not very proficient with C++ so I let me try to see a bit more what's the exact reason. 

#95 looks interesting and would be super-useful. Let me see if there's anything I can contribute there

**Comment by lemire** [comment]
If you follow the following pattern...

```
// use whatever means you want to get a string of your JSON document
std::string_view p = get_corpus(filename); // you are responsible for freeing p.data()
ParsedJson pj;
pj.allocateCapacity(p.size()); // allocate memory for parsing up to p.size() bytes
const int res = json_parse(p, pj); // do the parsing, return 0 on success
// parsing is done!
if (res != 0) {
    // You can use the "simdjson/simdjson.h" header to access the error message
    std::cout << "Error parsing:" << simdjson::errorMsg(res) << std::endl;
}
// You can safely delete the string content
aligned_free((void*)p.data());
```

You should be able to tell whether `allocateCapacity` fails. I am guessing that it would be your problem.

**Comment by TkTech** [comment]
> I am not sure pysimdjson does selective parsing. I think it parses the whole thing.

Correct, it's still just using an unmodified simdjson under the hood. It does selective elevation of elements to Python objects when using `items()`, that's it. But that's also 95% of the overhead!

**Comment by tahasabih** [comment]
Yes it definitely fails on allocate memory:

```
Could not allocate memory
double free or corruption (out)
Aborted (core dumped)
```


I am running this on a 64bit EC2 machine with 32GB of memory.

**Comment by lemire** [comment]
@tahasabih 

Thanks. This makes sense.

Note, no matter how we skin this cat, the performance is always going limited if your architecture involves bring large volumes of data from memory to cache, then pushing them back from cache to memory (as a materialized DOM) and then bring them back to cache.

A reasonable solution for us might to implement a form of paging. We start parsing the large document, then we stop. We let the user do what they want to do with the results... Then we resume parsing, effectively clearing the previously parsed data. And so forth. How to make this work nicely is an open question... but probably not impossibly hard.

**Comment by jpike88** [comment]
Why on earth would anyone generate JSON files that exceed 20/50/100mb, let alone 4GB? What DB is that, and why do you have to do it this way? This sounds more like a design issue to me.

**Comment by nichtich** [comment]
One real-world use case is [Wikidata dumps](https://www.wikidata.org/wiki/Wikidata:Database_download#JSON_dumps_(recommended)) with 36 GB gzipped. Each dump is one large JSON array that happens to contain one array member per line, so existing tools break up the array before parsing. In the end this data is used similar newline delimited JSON.

**Comment by lemire** [comment]
@nichtich  Nice example.

**Comment by spansh** [comment]
Also things like https://www.edsm.net/en/nightly-dumps I'm currently using the rapidjson incremental parsing mode but am very interested in improving that (That said I am very much at the long tail of improvements).

**Comment by lemire** [comment]
@spansh 

I love the link https://www.edsm.net/en/nightly-dumps

**Comment by tahasabih** [comment]
Firebase backups are also single JSON objects, so for large databases, it can easily exceed 10-20 GBs. This in itself is not an issue - we actually ended up going with gson streaming reads - it's extremely robust and memory efficient (albeit not as fast as simd).

**Comment by lemire** [comment]
@tahasabih  Thanks for another example.

I think you will agree that the goal is not to load tens of GBs in memory at once. We will have to go with some form of paging if we want to have good performance.

**Comment by tahasabih** [comment]
Yes definitely, paging or streaming reads would be really useful!

**Comment by effolkronium** [comment]
How about memory mapping for a super-large json files?. The OS will be able to handle memory usage by its own.

**Comment by lemire** [comment]
Another candidate https://www.yelp.com/dataset/challenge

**Comment by jkeiser** [comment]
@tahasabih how often is your file updated, and how often do you need to grab pieces of it? I'm trying to get a feel for whether a lightweight jsonpointer/jsonpath index would be a fit. The issue here is, no matter what, we have to scan the file to find your path, and we have to do a lot of the parsing work even if we optimize for finding a specific key. Like, we can't do use a `grep` algorithm to scan for something like `"abc"`, because we need to know whether it's the top level `"abc"`, and that means we at *least* need to count `{` and `}`, which means we need to parse strings so we know whether `{` and `}` are in string, and so on, and so on ...

That said, I've been thinking a lot lately about caller-driven parsing, where you tell the parser what you want, and what types you want it to be, and it ignores everything else as much as possible. I think there's potential there for sure.

**Comment by spansh** [comment]
In my case I need to parse the whole file and I largely want every field (or some subset of them).

Currently I'm using the SAX parser from RapidJSON which works how I'd like it, I've written my own sub parser using perfect hashes to direct things into appropriate struct fields.  I'm happy enough with that.

Ultimately (and my use case might be massively specialised to me).  I have a JSON array of objects (deep in this case so there are multiple levels within each object) and I'd like to process those objects one at a time.  The file is far too big to keep in memory so there's no option but to parse it in pieces.  I run the whole file at least once every 2 days, ideally every day.

That said the code I have works and optimising the JSON parsing aspect of it will give limited performance increase due to the other processing I have to do after I have a particular object.  My current c++ code essentially runs through the file and spits out bulk elasticsearch insert commands which I then collect into suitable chunks and throw at elasticsearch in parallel (limiting the number of parallel sends to some sensible number so as not to overload the cluster).  Decoupling elasticsearch I can process the file at 10,000 or so objects per second.  It drops to 2000 or so when I add in the elasticsearch side.  This is why I'm somewhat interested but I doubt it'll give me much of a speed increase (the file is actually bzip2'd and I read it through boost::iostreams with the bzip2 decompressor, it reads that file at around 3MB/second through inspection via "pv -d")

I do have other use cases where parsing the same file object by object (as a special case of a json array at the top level of objects wanted one by one) only wants a few sub fields.  I'd be hesitant to spend the time writing a special use case for that unless it turns out that everyone parsing these giant files wants things the same way, object by object.  A SAX parser would be suitable for everyone even though it would require people to write more code of their own.

**Comment by jkeiser** [comment]
@spansh if your code is public, would you mind linking it? When we think about the API, it'll be useful to look at real examples to see how various APIs would work.

> I'd be hesitant to spend the time writing a special use case for that unless it turns out that everyone parsing these giant files wants things the same way, object by object. A SAX parser would be suitable for everyone even though it would require people to write more code of their own.

Agreed that we should have a streaming API, particularly for people that already know SAX, and that's in the plans for sure.

But SAX is a real bear to code against in my experience, and I am starting to think there may be an easier API to handle forward-only stream parsing that gives equal or better performance. We shall see, though, and it won't block SAX :)

**Comment by spansh** [comment]
None of the code is currently public though the only thing from preventing that is my personal shame about how bad the code is (and it's tied in with a lot of other code).

I have a much smaller example (which is still arguably terrible code) but is filling a spatial database with a much smaller subset (and doesn't have 30+ gperf perfect hash files accompanying it) I could share, totalling maybe 1000 lines of code in this case.  Give me a little bit to get it onto github.

**Comment by spansh** [comment]
You can check out a version of that here: https://github.com/spansh/galaxy-spatial.  I decoupled Postgres and made it so it just generates insert statements on stderr instead.  Give me a shout if you have trouble building or understanding anything.  A lot of the code here deals with the SAX side in an attempt to be as efficient as possible, ultimately if I could get mini "documents" one at a time from the stream the code would be substantially simpler.

**Comment by lemire** [comment]
> ultimately if I could get mini "documents" one at a time from the stream the code would be substantially simpler

Interesting. Can you elaborate?

**Comment by spansh** [comment]
So where with simdjson you currently ```auto [doc, error] = document::parse(string("[ 1, 2, 3 ]"));``` and with the streaming api you do ```    simdjson::padded_string p = simdjson::get_corpus(filename);
    simdjson::ParsedJson pj;
    simdjson::JsonStream js{p};
    int parse_res = simdjson::SUCCESS_AND_HAS_MORE;
    
    while (parse_res == simdjson::SUCCESS_AND_HAS_MORE) {
            parse_res = js.json_parse(pj);

            //Do something with pj...
        }```

I'd essentially like something which works very similar to the jsonlines parser, so basically ignoring the first [ and the last ] (I could potentially accomplish the same myself by seeking the filehandle past the first [) then treating the entire document as a series of objects.  This does feel like a very special case but I get the impression that all of these enormous json files tend to be arrays of objects so it might be the main use case when parsing larger json files.

That way I have the DOM object and I can just pull out the fields I'm interested in via json pointers.  That or just loop through the actual object as described in the "Navigating the parsed document".  That would probably end up with a similar code to what I posted with the perfect hashes to compare keys, keeping track of level and comparisons of types.

**Comment by lemire** [comment]
@spansh That would be better (for the users) than SAX parsing. And we could potentially parallelize the parsing.

@jkeiser : I think that the proper way to do this would be to generalize the JsonStream. I will open an issue.

**Comment by jkeiser** [comment]
I suspect we can do even better, parsing less stuff when the user tells us what they need, but generalized incremental parsing is the absolute right first step, and good in its own right. :+1:

**Comment by jkeiser** [comment]
We have gone a long way towards efficiency with ondemand, allowing you to skip parsing data you don't use. However, we still require you to load the whole document as input.

While input streaming is very important, I'm leaning towards pushing it out to 2.0 and push out what we have into more users' hands--6GB is a real use case, and one we need to focus on (and have made progress towards), but it can certainly headline a 2.0 or 3.0 release.

**Comment by lemire** [comment]
Marked for 2.0.

**Comment by david542542** [comment]
Hi @lemire any update on this item? Or rather, do we have an estimate for when 2.0 will come out? We're currently using RapidJSON as it's the only one that can support streaming on large files (have tried simdjson and yyjson as well).

**Comment by lemire** [comment]
This was tagged with version 2.0, but version 2.0 has been published. We did not follow the milestones, which was a leftover project management technique. I renamed the old milestones, though they probably should just be deleted.


@david542542 If you can elaborate on your need, we could see what we could do to help with it. What I do not plan on doing is making it easy to just load very large monolithic JSON documents. I would rather discourage such uses. We want people to stream through their JSON and load it up in a database if they need to... not hold the big thing in a memory blob.

**Comment by spansh** [comment]
I did look at this again a few months ago and attempted to get simdjson to process my files it seemed like I'd have to do a lot of preprocessing to allow simdjson to handle this.  iterate_many did looked promising but it seems specifically targeted at ndjson. The files I'm processing are large json arrays.  I'm personally interested in each object one after another and in my main specific case there are newlines after each object but ideally those wouldn't be necessary.

Ultimately, I have a compressed 55GB file on disk, and I'm currently streaming that decompression and passing the stream to rapidjson and processing that with the SAX parser.

Since the file is compressed I can't simply pass the file to simdjson, though I can decompress it using a pipe rather than an in process decompression (so STDIN would be the file of uncompressed JSON).  Then I essentially want each object in the JSON array in turn.  I have several use cases for this, some of them I do need all of the attributes within each object, and others I simply need a small subset of each object.

```json
[
    {"id64":752911,"name":"Dryoea Flyou AA-A h0","coords":{"x":-35336.34375,"y":-220.0625,"z":18609.0625},"allegiance":null,"government":"None","primaryEconomy":"None","secondaryEconomy":"None","security":"Anarchy","population":0,"date":"2020-09-28 23:12:05+00","bodies":[{"id64":752911,"bodyId":0,"name":"Dryoea Flyou AA-A h0","type":"Star","subType":"O (Blue-White) Star","distanceToArrival":0.0,"mainStar":true,"age":2,"spectralClass":"O0","luminosity":"Vz","absoluteMagnitude":-8.754044,"solarMasses":30.921875,"solarRadius":8.42919918332135,"surfaceTemperature":45390.0,"rotationalPeriod":3.11085811966435,"axialTilt":0.0,"stations":[],"updateTime":"2020-09-28 23:12:11+00"},{"id64":360287970190392591,"bodyId":10,"name":"Dryoea Flyou AA-A h0 1","type":"Planet","subType":"Class III gas giant","distanceToArrival":8717.431244,"isLandable":false,"gravity":2.5615977363108,"earthMasses":286.302216,"radius":67397.456,"surfaceTemperature":658.69928,"surfacePressure":0.0,"atmosphereType":null,"parents":[{"Star":0}],"atmosphereComposition":{"Helium":24.551804,"Hydrogen":75.448204},"terraformingState":"Not terraformable","rotationalPeriod":1.92144769826389,"rotationalPeriodTidallyLocked":false,"axialTilt":-0.005029,"orbitalPeriod":3730.45527548701,"semiMajorAxis":14.8149309807504,"orbitalEccentricity":0.678556,"orbitalInclination":135.908962,"argOfPeriapsis":51.272645,"stations":[],"updateTime":"2020-09-28 23:12:28+00"}],"stations":[]}
]
```

The stations arrays are sometimes populated as well.  Anyways in some circumstances I'm only interested in the id64, x, y, z and name.  So having an optimised way of getting just those fields would be a bonus (otherwise I can simply have my code ignore them easily enough).  The full file is from https://spansh.co.uk/dumps and is https://downloads.spansh.co.uk/galaxy.json.gz.  It's 55GB compressed and I believe around 400GB uncompressed.  With rapidjson I can process the entire file in about 25 minutes (around 50,000 objects per second).  Process in this case means parsing the json, extracting a few of the fields and storing them in a struct which is then appended to a vector.

I have other use cases which parse the entire file and convert each one into a json format suitable to throw at a few elasticsearch indexes and then throws that at elasticsearch (that process takes a little over 3 hours).

**Comment by lemire** [comment]
> I'm confused. This thread has many examples of JSON files larger than 4GB, in which there is an array (either at the top-level or nested) whose elements the user wants to read using simdjson.

Yes. And that's something we could practically support with relative ease and high performance. The reason it is not currently supported is mostly a matter of user interface and format definitions.

**Comment by effolkronium** [comment]
Just use mapped files and pass pointer to the data to the json lib

**Comment by lemire** [comment]
I am currently thinking that we could extend the JSON streaming functionality (which supports infinite streams of documents) to JSON arrays. People who have large arrays embedded in complex JSON documents would be responsible for locating the start of the array and pointing us at it. This would immediately meet @spansh needs.

Note that the `iterate_many` functionality would precisely address the use case they describe.

**Comment by david542542** [comment]
@lemire we have a lot of user-generated json documents (where we cannot pre-process), with the only guarantee that the document is valid JSON and we don't want to have to say to the user if they upload a 4GB file:

> "In practice, your file doesn't exist because a large file is always a stream of small JSON documents and not one big JSON file. Please, load an existing file."

Obviously I say that tongue in cheek, but essentially our need is just that we have large json files of various formats and want to be able to use simdjson (if possible) instead of rapidjson, as hopefully the former is much faster at parsing large documents.

**Comment by lemire** [comment]
@spansh I have downloaded your large compressed JSON file. I won't work on this right now, but soon I will use it for benchmarking.

**Comment by david542542** [comment]
@lemire sounds good, let me see what kind of effort might be involved to contribute with the said feature/improvement.

**Comment by lemire** [comment]
@david542542 

> let me see what kind of effort might be involved to contribute with the said feature/improvement.

One will likely require a new simdjson front-end. For example, the current code base uses 32-bit counters for string lengths which limits the length of any given string to 4GB, once you allow individual strings to exceed 4GB, then you must switch to 64-bit counters. And that's the least of it. The current design is based on a fixed capacity so that your memory usage is fixed over time, no matter how many documents you access... such a design is no longer applicable if you allow unbounded memory usage. Dealing with potentially infinite memory usage and the related error conditions (e.g., page faults) will require in-depth testing. In effect, your use case is not one that we cover: potentially infinitely large monolithic JSON documents. To address a whole new use case, there needs to be substantial engineering effort. I would estimate that an experienced C++ engineer would take at least six months, with a margin of error based on their expertise [The final performance is likely to be limited by paging and related issues](https://lemire.me/blog/2021/08/21/the-big-load-anti-pattern/).

**Comment by theogaraj** [comment]
Hello, I've been evaluating the Python bindings for `simdjson` (`pysimdjson`) and am interested in a "streaming" capability.  I posted a [question](https://github.com/TkTech/pysimdjson/issues/120) in the `pysimdjson` repo and was directed here by @TkTech.
 
I have large (4GB compressed, ~ 30GB uncompressed) JSON documents that contain (among other things) a couple of arrays, each of which contain many individual JSON objects which I want to process one by one.  (Yes, it would be much better if these were jsonlines/ndjson, but I don't control the file creation.)  I am currently using [jsonslicer](https://pypi.org/project/jsonslicer/) which is based on [yajl](https://lloyd.github.io/yajl/).  It does exactly what I need, but I'm evaluating other options for faster performance.

From https://github.com/simdjson/simdjson/issues/128#issuecomment-1171297063 :
> I am currently thinking that we could extend the JSON streaming functionality (which supports infinite streams of documents) to JSON arrays. People who have large arrays embedded in complex JSON documents would be responsible for locating the start of the array and pointing us at it. This would immediately meet @spansh needs.
Note that the iterate_many functionality would precisely address the use case they describe.

☝️ This sounds like what I need, but it's not clear whether this already exists or is still just an idea.  May I get some guidance/clarification please?

**Comment by lemire** [comment]
@theogaraj 

Currently, we support huge/infinite files with `jsonlines/ndjson`. We do not currently support files larger than 4 GB with our main API. 

> This sounds like what I need, but it's not clear whether this already exists or is still just an idea. May I get some guidance/clarification please?


Processing 30 GB monolithic JSON files with high performance is an interesting but specific problem. We do invite contributions.

**Comment by lemire** [comment]
The main use case in this thread, a huge file that is really one big JSON array of records, is now covered. Since #2674, `iterate_many` (On Demand) and `parse_many` (DOM) accept `simdjson::stream_format::comma_delimited_array`: the parser strips the outer `[`/`]` and streams the elements one at a time, in batches, so memory use stays bounded no matter how large the file is. Combine it with `simdjson::padded_memory_map` so the file is memory-mapped rather than loaded:

```cpp
simdjson::padded_memory_map map("huge.json");
ondemand::parser parser;
ondemand::document_stream stream;
auto error = parser.iterate_many(map.view(), ondemand::DEFAULT_BATCH_SIZE,
                                 simdjson::stream_format::comma_delimited_array).get(stream);
for (auto doc : stream) { /* one array element at a time */ }
```

See "JSON Array As A Document Stream" in `doc/iterate_many.md`. This is on `master` and will ship in the next release (it is not in v4.6.11).

We still don't plan to support single monolithic documents over 4 GB (see the discussion above), so I am closing this. If you have a concrete need this doesn't cover, please open a new, focused issue.
